lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,191 @@
+
+	// maxBufferPerRequest limits individual request body size to allow
+	// balanced memory usage across requests. Kubernetes ConfigMaps and Secrets
+	// have 1 MiB etcd limits, so 50 MiB is generous.
+	maxBufferPerRequest = 50 * 1024 * 1024 // 50 MiB
+)
+
+var retryBufferSemaphore = semaphore.NewWeighted(maxTotalBufferSize)
+
+// newRetryableTransport creates a retryableTransport that wraps the provided
+// HTTP/2 transport to enable automatic retry on GOAWAY frames.
+//
+// Intended to wrap transports created via newH2Transport() or equivalent
+// HTTP/2-configured transports. Wrapping non-HTTP/2 transports would add
+// unnecessary buffering.
+func newRetryableTransport(inner http.RoundTripper, log *slog.Logger) *retryableTransport {
+	return &retryableTransport{
+		inner:     inner,
+		log:       log,
+		semaphore: retryBufferSemaphore,
+	}
+}
+
+// RoundTrip implements http.RoundTripper and makes requests retryable.
+func (rt *retryableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Detect HTTP/1.1 protocol upgrades that won't receive HTTP/2 GOAWAY.
+	if rt.isUpgrade(req) {
+		// Perform roundtrip, no retry available.
+		return rt.inner.RoundTrip(req)
+	}
+
+	if req.Body == nil || req.GetBody != nil {
+		// Perform roundtrip, retry available.
+		return rt.inner.RoundTrip(req)
+	}
+
+	// Buffer body to make retryable.
+	releaseSemaphore, err := rt.makeRetryable(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if releaseSemaphore != nil {
+		// Release semaphore after roundtrip completes.
+		defer releaseSemaphore()
+	}
+
+	// Perform roundtrip, retry available.
+	return rt.inner.RoundTrip(req)
+}
+
+// makeRetryable buffers the request body to enable retry on HTTP/2 GOAWAY.
+func (rt *retryableTransport) makeRetryable(req *http.Request) (func(), error) {
+	// Chunked request are not retried due to unknown size.
+	if req.ContentLength < 0 {
+		rt.log.DebugContext(req.Context(),
+			"Skipping retry buffer for chunked encoding, request will not be retryable on GOAWAY",
+			"method", req.Method,
+			"path", req.URL.Path)
+		return nil, nil
+	}
+
+	// Request body is too large to retry.
+	// Allows balanced use of memory budget across requests.
+	if req.ContentLength > maxBufferPerRequest {
+		rt.log.InfoContext(req.Context(),
+			"Request body too large for retry buffer, request will not be retryable on GOAWAY",
+			"size", req.ContentLength,
+			"limit", maxBufferPerRequest,
+			"method", req.Method,
+			"path", req.URL.Path)
+		return nil, nil
+	}
+
+	// Acquire memory, blocks if unavailable.
+	if err := rt.semaphore.Acquire(req.Context(), req.ContentLength); err != nil {
+		rt.log.InfoContext(req.Context(),
+			"Unable to acquire retry buffer semaphore, request will not be retryable on GOAWAY",
+			"size", req.ContentLength,
+			"method", req.Method,
+			"path", req.URL.Path,
+			"error", err)
+		return nil, nil
+	}
+	releaseSemaphore := func() { rt.semaphore.Release(req.ContentLength) }
```

Comment:
We don't care about releasing just the semaphore, we can release the semaphore only after the memory is actually released.